### ISSUE TITLE
Make BaseRecord class

### ISSUE THREAD

**Issue by sondreso** [equinor/ert#1792]
The current implementation of the record is becoming complicated. We should make BaseRecord class with BlobRecord and NumericalRecords as derived classes.

**Comment by verveerpj** [comment]
Working on this seems premature, blob records are still in progress, and I am assuming the current Record* types would be numerical records. The latter are all basically pydantic classes, with a common base, that do not seem to expose any methods, i.e. basically just validated data structures. It seems we are still discussion how to do those (#1731 ).

Does it even make sense at this point to design a base class?

@markusdregi  what do you think?

**Comment by markusdregi** [comment]
Actually I don't think #1731 will do that much to the `ert.data.Record` implementation. It will probably just use the `LIST_BYTES` type that is already present. I think it would make sense to create an abstract `Record` with some implementation and some abstract functions high lighting the behaviour of any `Record` and then introduce a `BlobRecord` that deals with the `LIST_BYTES` type and a `NumericalRecord` that deals with the other current types supported...

**Comment by verveerpj** [comment]
So, then the current `Record` would become `NumericalRecord`, and we should remove the `LIST_BYTES` type from that when the `BlobRecord` is implemented? 

And another question: are `EnsembleRecord` and `MultiEnsembleRecord` really records that should derived from `BaseRecord`? They seem to be a bit of a different kind of beast, more collections than data records. Transmitters do  not seem to support them, while` ert_storage` does. According to their names they should derive from a `BaseRecord`, but I guess the question is what really the base functionality of a record should be. Is it something that can be transmitted and stored? If so, should the ensemble records support that?

**Comment by markusdregi** [comment]
I would suggest that we end up with two concrete records, being `NumericalRecord` and `BlobRecord` that together form the union of data that the current `Record` can carry. Then, we introduce a `BaseRecord` that is an (abstract?) class both `NumericalRecord` and `BlobRecord` inherits and the idea would be that most of ERT would (especially the data transportation and workflow pars) should not really know what kind of record is being passed around. In particular, unless you need the numeric content you should type and treat it like a `BaseRecord`.

Then, since most of the application will be passing this base record around. Should it be named `BaseRecord` or just `Record` ?  🤷🏻 

Last, you are right. `EnsembleRecord` and `MultiEnsembleRecord` are containers for records. They should be able to carry both `Blob` and `Numerical` records, but beyond they should not derive from the base record themselves...

**Comment by verveerpj** [comment]
Yes, I was already starting along these lines. I renamed the current `Record` class to `NumericalRecord` and started coding an abstract `Record` class. This would have all the (possibly concrete) methods and properties needed to handle common functionality. 

I picked `Record` not `BaseRecord` since from the point of view of much of the code, it should be a generic 'record' object, in principle there is no need to know if it is a base class or some duck-typed thingy (but the base class is handy for static typing and runtime type checking and such). The record transmitters do this already: there is an abstract base class `RecordTransmitter`.

The names `EnsembleRecord` and `MultiEnsembleRecord` are, in retrospect, a bit unlucky since they suggest that they are records. However, now they will definitely not be. Should we consider renaming?
